### PR TITLE
Denominator CPU code

### PR DESCRIPTION
**Pull request by vimalmanohar**
@YiwenShaoStephen I wrote the denominator computation in CPU and made some changes to the denominator graph for the same. See if you did something similar and compare the two.

### CODE REVIEW

**Comment by YiwenShaoStephen** [comment]
@vimalmanohar Yes, I check it and I think basically they are the same. You can merge it. Thanks!

**Comment by YiwenShaoStephen** [comment]
@vimalmanohar Is it OK to merge it now? I made some small modifications on it thus to make the following forward and backward function in pytorch more convenient. And once this PR is merged, I can make another PR to show how loss function can be defined and used.
